### PR TITLE
Feature/backup restore api based python code

### PR DESCRIPTION
**Pull request by ashviniwaghmode**
Open source the single and group VM backup and restore API code

### CODE REVIEW

File: recipes/python/backup-restore/common.py [hunk 1]
```diff
@@ -0,0 +1,271 @@
...
+# Create protection plan
+def create_protection_plan(baseurl, token, protection_plan_name, storage_unit_name):
+    """ This function create the protection plan """
+    print(f"Create protection plan:[{protection_plan_name}]")
+    headers.update({'Authorization': token})
+    url = f"{baseurl}servicecatalog/slos?meta=accessControlId"
+
+    cur_dir = os.path.dirname(os.path.abspath(__file__))
+    file_name = os.path.join(cur_dir, "create_protection_plan_template.json")
+    file_handle = open(file_name)
```
Comment (rkennedy) [inline recipes/python/backup-restore/common.py]: Use `with` to ensure the file is closed promptly.

```suggestion
    with open(file_name, 'r') as file_handle:
```

File: recipes/python/backup-restore/common.py [hunk 2]
```diff
@@ -0,0 +1,271 @@
...
+
+import json
+import os
+import sys
+import time
+import requests
+
+headers = {"Content-Type" : "application/vnd.netbackup+json;version=4.0"}
+
+# Get the base NetBackup url
```
Comment (rkennedy) [inline recipes/python/backup-restore/common.py:15]: These comments before each function definition contribute nothing. They're essentially just the function name without the underscores. The docstrings and function names ought to be sufficient descriptions of the functions.

File: recipes/python/backup-restore/README.md [hunk 3]
```diff
@@ -0,0 +1,130 @@
+# Netbackup VmWare agentless single and group VM backup and restore APIs code samples
```
Comment (rkennedy) [inline recipes/python/backup-restore/README.md]:
```suggestion
# NetBackup VmWare agentless single and group VM backup and restore APIs code samples
```

File: recipes/python/backup-restore/common.py (hunk 1 above)
Comment (ashviniwaghmode) [reply recipes/python/backup-restore/common.py]: Used 'with' statement to open the file.

File: recipes/python/backup-restore/common.py (hunk 2 above)
Comment (ashviniwaghmode) [reply recipes/python/backup-restore/common.py:15]: In this common.py file, have small functions so described according to that only.

Comment (ashviniwaghmode) [reply recipes/python/backup-restore/common.py:15]: In this common.py file, have small functions so described according to that only.

Comment (ashviniwaghmode) [reply recipes/python/backup-restore/common.py:15]: In this common.py file, have small functions so described according to that only.

File: recipes/python/backup-restore/README.md (hunk 3 above)
Comment (ashviniwaghmode) [reply recipes/python/backup-restore/README.md]: Changed NetBackup to NetBackup

File: recipes/python/backup-restore/common.py (hunk 2 above)
Comment (rkennedy) [reply recipes/python/backup-restore/common.py:15]: I don't understand that response.

**Review by rkennedy — DISMISSED** [review]

File: recipes/python/backup-restore/README.md [hunk 4]
```diff
@@ -0,0 +1,130 @@
+# NetBackup VmWare agentless single and group VM backup and restore APIs code samples
+
+## Executing the scripts:
+
+Pre-requisites:
+- NetBackup 8.2.1 or higher
```
Comment (santhosh-vrts) [inline recipes/python/backup-restore/README.md]:
```suggestion
- NetBackup 8.3 or higher      
```

File: recipes/python/backup-restore/README.md [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+
+Pre-requisites:
+- NetBackup 8.2.1 or higher      
+- Python 3.6 or higher
+- Python modules: `requests`
+
+Use the following commands to run the scripts.
+### - Single VM backup and restore
+
+This single_vm_backup_restore.py script describes how to backup VM using protection plan and restore the single VMware virtual machine that were backuped with NetBackup for VMware.
```
Comment (santhosh-vrts) [inline recipes/python/backup-restore/README.md]:
```suggestion
This single_vm_backup_restore.py script demonstrates how to backup a VM (VMware virtual machine) using a protection plan and restore it using NetBackup APIs.
```

File: recipes/python/backup-restore/README.md [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+- Subscribe asset to protection plan
+- Perform immediate backup
+- Verify the status of jobs
+- Initiate instant access recovery
+- Verify state of instant access recovery VM.
+- Perform the cleanup(e.g. remove instant access VM, subscription, protection plan and vcenter)
+
+### - Group VM backup and restore
+
+This group_vm_backup_restore.py script describes how to backup multiple VMs using protection plan and perform bulk restore of VMware virtual machines that were backuped with NetBackup for VMware.
```
Comment (santhosh-vrts) [inline recipes/python/backup-restore/README.md]:
```suggestion
This group_vm_backup_restore.py script demonstrates how to backup multiple VMs (VMware virtual machines) using a protection plan and perform bulk restore of the VMs using NetBackup APIs.
```

File: recipes/python/backup-restore/README.md (hunk 4 above)
Comment (ashviniwaghmode) [reply recipes/python/backup-restore/README.md]: Changed the NetBackup version to 8.3

File: recipes/python/backup-restore/README.md (hunk 5 above)
Comment (ashviniwaghmode) [reply recipes/python/backup-restore/README.md]: Changed the description

Comment (ashviniwaghmode) [reply recipes/python/backup-restore/README.md]: Changed the description

Comment (ashviniwaghmode) [reply recipes/python/backup-restore/README.md]: Changed the description

File: recipes/python/backup-restore/README.md (hunk 6 above)
Comment (ashviniwaghmode) [reply recipes/python/backup-restore/README.md]: Changed the description

File: recipes/python/backup-restore/README.md [hunk 7]
```diff
@@ -0,0 +1,130 @@
+# NetBackup VmWare agentless single and group VM backup and restore APIs code samples
```
Comment (jaya-korlepara) [inline recipes/python/backup-restore/README.md]:
```suggestion
# NetBackup VMware agentless single and group VM backup and restore APIs code samples
```

File: recipes/python/backup-restore/README.md [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+  --master_server MASTER_SERVER
+                        NetBackup master server name
+  --master_server_port MASTER_SERVER_PORT
+                        NetBackup master server port
+  --master_username MASTER_USERNAME
+                        NetBackup master server username
+  --master_password MASTER_PASSWORD
+                        NetBackup master server password
+  --vcenter_name VCENTER_NAME
+                        Vcenter name
```
Comment (jaya-korlepara) [inline recipes/python/backup-restore/README.md]:
```suggestion
                        vCenter name
```

File: recipes/python/backup-restore/README.md [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+  --vcenter_username VCENTER_USERNAME
+                        Vcenter username
```
Comment (jaya-korlepara) [inline recipes/python/backup-restore/README.md]:
```suggestion
                        vCenter username
```

File: recipes/python/backup-restore/README.md [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+  --vcenter_password VCENTER_PASSWORD
+                        Vcenter password
```
Comment (jaya-korlepara) [inline recipes/python/backup-restore/README.md]:
```suggestion
                        vCenter password
```

File: recipes/python/backup-restore/README.md [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+  --vcenter_port VCENTER_PORT
+                        Vcenter port
```
Comment (jaya-korlepara) [inline recipes/python/backup-restore/README.md]:
```suggestion
                        vCenter port
```

File: recipes/python/backup-restore/README.md [hunk 12]
```diff
@@ -0,0 +1,130 @@
...
+  --protection_plan_name PROTECTION_PLAN_NAME
+                        Protection plan name
+  --clientvm CLIENTVM   Client VM name
+  --restore_vmname RESTORE_VMNAME
+                        Restore VM name
+```
+
+Execution flow of single VM backup and restore script:
+- Add vcenter to NBU master
+- Discovery of Vcenter
```
Comment (jaya-korlepara) [inline recipes/python/backup-restore/README.md:65]:
```suggestion
- Discovery of vCenter
```

File: recipes/python/backup-restore/README.md [hunk 13]
```diff
@@ -0,0 +1,130 @@
...
+  --master_server MASTER_SERVER
+                        NetBackup master server
+  --master_server_port MASTER_SERVER_PORT
+                        NetBackup port
+  --master_username MASTER_USERNAME
+                        NetBackup master server user name
+  --master_password MASTER_PASSWORD
+                        NetBackup master server password
+  --vcenter_name VCENTER_NAME
+                        Vcenter name
```
Comment (jaya-korlepara) [inline recipes/python/backup-restore/README.md:110]:
```suggestion
                        vCenter name
```
Please replace  "Vcenter" with "vCenter" in rest of the file.

File: recipes/python/backup-restore/common.py [hunk 14]
```diff
@@ -0,0 +1,271 @@
...
+    print(f"Add the vCenter credential:[{vcenter_server}]")
+    headers.update({'Authorization': token})
+    url = f"{baseurl}config/servers/vmservers"
+    payload = {'serverName':vcenter_server, 'vmType':vcenter_server_type, 'userId':vcenter_username,\
+                    'password':vcenter_password, 'port':vcenter_port}
+    status_code, response_text = rest_request('POST', url, headers, data=payload)
+    validate_response(status_code, 201, response_text)
+    print(f"vCenter credentials added successfully:[{vcenter_server}]")
+
+# Get Vmware server discovery status
```
Comment (jaya-korlepara) [inline recipes/python/backup-restore/common.py:44]:
```suggestion
# Get VMware server discovery status
```

File: recipes/python/backup-restore/common.py [hunk 15]
```diff
@@ -0,0 +1,271 @@
...
+    uuid = response_text['data'][0]['attributes']['instanceUuid']
+    exsi_host = response_text['data'][0]['attributes']['host']
+
+    print(f"Client asset Id:[{asset_id}]")
+    print(f"Client uuid Id:[{uuid}]")
+    print(f"Client exsi host:[{exsi_host}]")
+    return asset_id, uuid, exsi_host
+
+# Get StorageUnits
+def get_storage_units(baseurl, token):
```
Comment (jaya-korlepara) [inline recipes/python/backup-restore/common.py]: There could be multiple storage units configured. This function checks the capability of the first storage unit and fails if it does not support instant access.

Can this functiton take the name of the stu as an input instead?

File: recipes/python/backup-restore/common.py [hunk 16]
```diff
@@ -0,0 +1,271 @@
...
+    """ This function return the storage unit name """
+    headers.update({'Authorization': token})
+    url = f"{baseurl}storage/storage-units"
+    status_code, response_text = rest_request('GET', url, headers)
+    validate_response(status_code, 200, response_text)
+    storage_unit_name = response_text['data'][0]['id']
```
Comment (jaya-korlepara) [inline recipes/python/backup-restore/common.py]: Will this work if there are no storage units configured?

File: recipes/python/backup-restore/common.py [hunk 17]
```diff
@@ -0,0 +1,271 @@
...
+
+# Create protection plan
+def create_protection_plan(baseurl, token, protection_plan_name, storage_unit_name):
+    """ This function create the protection plan """
+    print(f"Create protection plan:[{protection_plan_name}]")
+    headers.update({'Authorization': token})
+    url = f"{baseurl}servicecatalog/slos?meta=accessControlId"
+
+    cur_dir = os.path.dirname(os.path.abspath(__file__))
+    file_name = os.path.join(cur_dir, "create_protection_plan_template.json")
```
Comment (jaya-korlepara) [inline recipes/python/backup-restore/common.py:114]: Could you describe the create_protection_plan_template.json in the Readme? May be add a note that this needs to exist as a prerequisite?

File: recipes/python/backup-restore/common.py [hunk 18]
```diff
@@ -0,0 +1,271 @@
...
+                print(f"Job:[{jobid}] completed with expected state:[{expected_state}]")
+                status = response_text['data']['attributes']['status']
+                print(f"Actual status:[{status}] and expected status:[{expected_status}]")
+                if status == expected_status:
+                    print(f"Job:[{jobid}] completed with expected status:[{expected_status}]")
+                    break
+                else:
+                    print(f"Failed backup jobid:[{jobid}] with status:[{status}]")
+                    raise Exception(f"Failed backup jobid:[{jobid}] with status:[{status}]")
+        else:
```
Comment (jaya-korlepara) [inline recipes/python/backup-restore/common.py:183]: indentation?

File: recipes/python/backup-restore/README.md [hunk 19]
```diff
@@ -0,0 +1,130 @@
...
+
+Pre-requisites:
+- NetBackup 8.3 or higher
+- Python 3.6 or higher
+- Python modules: `requests`
+
+Use the following commands to run the scripts.
+### - Single VM backup and restore
+
+This single_vm_backup_restore.py script demonstrates how to backup a VM (VMware virtual machine) using a protection plan and restore it using NetBackup APIs.
```
Comment (jaya-korlepara) [inline recipes/python/backup-restore/README.md]: Can you also mention that the restores are instant access restores?

File: recipes/python/backup-restore/common.py (hunk 18 above)
Comment (ashviniwaghmode) [reply recipes/python/backup-restore/common.py:183]: This is the while-else block. Indentation is correct here.

File: recipes/python/backup-restore/README.md (hunk 7 above)
Comment (ashviniwaghmode) [reply recipes/python/backup-restore/README.md]: Changed to VMware.

File: recipes/python/backup-restore/README.md (hunk 8 above)
Comment (ashviniwaghmode) [reply recipes/python/backup-restore/README.md]: Changed to vCenter.

File: recipes/python/backup-restore/README.md (hunk 9 above)
Comment (ashviniwaghmode) [reply recipes/python/backup-restore/README.md]: Changed to vCenter.

File: recipes/python/backup-restore/README.md (hunk 10 above)
Comment (ashviniwaghmode) [reply recipes/python/backup-restore/README.md]: Changed to vCenter.

File: recipes/python/backup-restore/README.md (hunk 11 above)
Comment (ashviniwaghmode) [reply recipes/python/backup-restore/README.md]: Changed to vCenter.

File: recipes/python/backup-restore/common.py (hunk 15 above)
Comment (ashviniwaghmode) [reply recipes/python/backup-restore/common.py]: Yes. Taken stu_name as input parameter and changed the function name to 'verify_stu_instant_access_enable'.

File: recipes/python/backup-restore/common.py (hunk 16 above)
Comment (ashviniwaghmode) [reply recipes/python/backup-restore/common.py]: Changed the approach. As per above comment.

File: recipes/python/backup-restore/common.py (hunk 17 above)
Comment (ashviniwaghmode) [reply recipes/python/backup-restore/common.py:114]: Added into the prerequisite.

File: recipes/python/backup-restore/README.md (hunk 19 above)
Comment (ashviniwaghmode) [reply recipes/python/backup-restore/README.md]: Done.

**Review by jaya-korlepara — APPROVED** [review]

**Pull request merged**